plugin/hierarchy.py

```diff
@@ -0,0 +1,211 @@
+        request1: Callable[..., Request],
+        request2: Callable[..., Request],
+        request_handler1: Callable[..., List[HierarchyItem]],
+        request_handler2: Callable[..., List[HierarchyItem]],
+        direction: int,
+        root_elements: List[HierarchyItem]
+    ) -> None:
+        self.weaksession = weaksession
+        self.request1 = request1
+        self.request2 = request2
+        self.request_handler1 = request_handler1
+        self.request_handler2 = request_handler2
+        self.direction = direction
+        self.root_elements = root_elements
+        session = self.weaksession()
+        self.session_name = session.config.name if session else None
+
+    def get_children(self, element: Optional[HierarchyItem]) -> Promise[List[HierarchyItem]]:
+        if element is None:
+            return Promise.resolve(self.root_elements)
+        session = self.weaksession()
+        if not session:
+            return Promise.resolve([])
+        if self.direction == 1:
+            return session.send_request_task(self.request1({'item': element})).then(self.request_handler1)
+        if self.direction == 2:
+            return session.send_request_task(self.request2({'item': element})).then(self.request_handler2)
+        return Promise.resolve([])
+
+    def get_tree_item(self, element: HierarchyItem) -> TreeItem:
+        command_url = sublime.command_url('lsp_open_location', {
+            'location': {
+                'targetUri': element['uri'],
+                'targetRange': element['range'],
+                'targetSelectionRange': element['selectionRange']
+            },
+            'session_name': self.session_name,
+            'flags': sublime.ADD_TO_SELECTION | sublime.SEMI_TRANSIENT | sublime.CLEAR_TO_RIGHT
+        })
+        path = simple_path(self.weaksession(), element['uri'])
+        return TreeItem(
+            element['name'],
+            kind=SYMBOL_KINDS.get(element['kind'], sublime.KIND_AMBIGUOUS),
+            description=element.get('detail', ""),
+            tooltip="{}:{}".format(path, element['selectionRange']['start']['line'] + 1),
+            command_url=command_url
+        )
+
+
+def make_data_provider(
+    weaksession: 'weakref.ref[Session]', sheet_name: str, direction: int, response: List[HierarchyItem]
+) -> HierarchyDataProvider:
+    if sheet_name == "Call Hierarchy":
+        request1 = Request.incomingCalls
+        request2 = Request.outgoingCalls
+        handler1 = lambda response: [incoming_call['from'] for incoming_call in response] if isinstance(response, list) else []  # noqa: E501,E731
+        handler2 = lambda response: [outgoing_call['to'] for outgoing_call in response] if isinstance(response, list) else []  # noqa: E501,E731
+    elif sheet_name == "Type Hierarchy":
+        request1 = Request.supertypes
+        request2 = Request.subtypes
+        handler1 = handler2 = lambda response: response if isinstance(response, list) else []  # noqa: E731
```

Comment:
Isn't it an instance of our `Error` in case of clangd? It would be more explicit to check that if that's the case.

Reply:
Yes, but it can also be `None` if it's no error. So this `isinstance(..., list)` catches all cases at once. That's basically why I used it here, so that it fits into a single line of code (and if I see it correctly, because the `send_request_task` doesn't have an anology to the `send_request_async`'s `on_error` callback).

Reply:
Should I create separate, named functions for the response handlers, instead of lambdas, to make it more explicit? (flake8 already complained :D)

Reply:
Yes, separate named functions would be better IMO.

Also those request1, request2, etc. names don't speak to me at all so I would perhaps suggest refactoring this way: 5c7fa1818a62be77c3e860db0d50fd04773fc670 (not sure if I got everything right but something like that).

Reply:
Yeah, makes more sense to do it that way. When you move things around and focus on initial ideas, you sometimes don't see the obvious.